test/engine/update.result

```diff
@@ -850,3 +850,287 @@ s:update({1}, {{'=', 'field3', {d = 30, e = 31, f = 32}}})
+    a positive integer'
+...
+s:update({1}, {{':', 3, 1, 0, '!!'}})
+---
+- error: 'Argument type in operation '':'' on field 3 does not match field type: expected
+    a string'
+...
+s:drop()
+---
+...
+-- Negative field number, fixed field_count
+format = {}
+---
+...
+format[1] = {name = 'field1', type = 'unsigned'}
+---
+...
+format[2] = {name = 'field2', type = 'unsigned'}
+---
+...
+format[3] = {name = 'field3', type = 'unsigned', is_nullable = true}
+---
+...
+format[4] = {name = 'field4', type = 'unsigned', is_nullable = true}
+---
+...
+s = box.schema.create_space('test', {format = format, field_count = 4})
+---
+...
+pk = s:create_index('pk')
+---
+...
+s:replace{1, 2, box.NULL, box.NULL}
+---
+- [1, 2, null, null]
+...
+s:update({1}, {{'!', -1, 42}})
+---
+- error: Tuple field count 5 does not match space field count 4
+...
+s:update({1}, {{'=', -1, 128}})
+---
+- [1, 2, null, 128]
+...
+s:drop()
+---
+...
+-- Negative field number, no field_count
+format = {}
+---
+...
+format[1] = {name = 'field1', type = 'unsigned'}
+---
+...
+format[2] = {name = 'field2', type = 'unsigned'}
+---
+...
+format[3] = {name = 'field3', type = 'unsigned', is_nullable = true}
+---
+...
+format[4] = {name = 'field4', type = 'unsigned', is_nullable = true}
+---
+...
+s = box.schema.create_space('test', {format = format})
+---
+...
+pk = s:create_index('pk')
+---
+...
+s:replace{1, 2}
+---
+- [1, 2]
+...
+s:update({1}, {{'!', -1, 42}})
+---
+- [1, 2, 42]
+...
+s:update({1}, {{'=', -1, 128}})
+---
+- [1, 2, 128]
+...
+s:drop()
+---
+...
+-- '#' doesn't trim nulls
+s = box.schema.create_space('test')
+---
+...
+pk = s:create_index('pk')
+---
+...
+s:replace{1, 2}
+---
+- [1, 2]
+...
+s:update({1}, {{'!', 4, 0}})
+---
+- [1, 2, null, 0]
+...
+s:update({1}, {{'#', 4, 1}})
+---
+- [1, 2, null]
+...
+s:drop()
+---
+...
+-- Maps (fail if don't exist)
+format = {}
+---
+...
+format[1] = {name = 'field1', type = 'unsigned'}
+---
+...
+format[2] = {name = 'field2', type = 'map'}
+---
+...
+s = box.schema.create_space('test', {format = format})
+---
+...
+pk = s:create_index('pk')
+---
+...
+map = {key1 = 1, key2 = 2}
+---
+...
+s:replace{1, map}
+---
+- [1, {'key1': 1, 'key2': 2}]
+...
+s:update({1}, {{'!', 'field42', 0}})
+---
+- error: Field 'field42' was not found in the tuple
+...
+s:update({1}, {{'!', '[3].key1', 1}})
+---
+- error: Field ''[3].key1'' was not found in the tuple
+...
+s:update({1}, {{'!', 3, 3}})
+---
+- [1, {'key1': 1, 'key2': 2}, 3]
+...
+s:drop()
+---
+...
+-- Arrays (fail if don't exist)
+format = {}
+---
+...
+format[1] = {name = 'field1', type = 'unsigned'}
+---
+...
+format[2] = {name = 'field2', type = 'array'}
+---
+...
+s = box.schema.create_space('test', {format = format})
+---
+...
+pk = s:create_index('pk')
+---
+...
+arr = {11, 22, {111, 222}}
+---
+...
+s:replace{1, arr}
+---
+- [1, [11, 22, [111, 222]]]
+...
+s:update({1}, {{'!', '[2][42]', 0}})
+---
+- error: Field ''[2][42]'' was not found in the tuple
```

Comment:
That does not seem right. Field `[2]` exists and it is an array. It means there should no be a reason why it can't auto-fill the fields until `[42]`. Tuple's root array is not any different from the nested arrays. At least this was the idea behind the xrow update code.

Also I realized that probably `#` should not throw an error. Because if the feature's main purpose is to fix the 'virtual' fields assignment and `=` will work on them now, then `#` should work as well. But now it fails. For example:
```
t = box.tuple.new({1, 2, 3})
t:update({{'#', 5, 1}})
```
This fails. But if I replace it with `=` or `!` - it works. That makes `#` and `!` not symmetric. Maybe such deletion should become nop when out of range or a key wasn't found (if it deletes from a map). I am afraid you need to ask Mons about that. 2 questions in total:
1) Do we auto-fill the nested arrays or only the root array?
2) Should we make `#` nop when it is out of range? To be symmetric with the new behaviour of `!`.

Reply:
1. I think we should auto-fill nested structures in every case, except type collision (update array field on map)
2. Agree with you, It's ok to update `#` also

Reply:
What do you mean as every case? For instance, I have a tuple `{1, 2, 3}`. And I make update `{{'=', '[5].key[3]', 100}}`. Should it also auto-create the intermediate array and map so it would result into this: `{1, 2, 3, nil, {key = {nil, nil, 100}}}`? Or is this an error, and we only auto-fill arrays and never create new arrays and maps on the path?

Reply:
@Mons, see above.

Reply:
This is an interesting question.
By example from Perl (called autovivification) I know, that feature may lead to unexpected results, but in most cases, it is rather handy.
But since we give an option to update the deep field and give no handy option to update such a field in case of it's absence, I consider we should support auto-creation of missing structures and raise an error only in case of conflicting types (array vs map)

Reply:
According to our last conversation with Mons and Alexander, we decided to make changes regarding auto-fill nested structures in a separate task

Reply:
What about the example I showed above about filling a non-root array? It does not need creation of the nested structures, the array exists already, it is just is not in the root. And yet it does not work now.

Reply:
This case will also be implemented in a separate task, according to our discussion. @Mons, @alyapunov  i am right?

Reply:
If you are not going to implement it now, it will work 50/50. For example, it will work if I will json-update some internal array using 2 operations instead of one. The second one would turn it from a bar into a rope in the code, and the auto-fill will work like in the root. But it won't work if the update operation is just one. Which looks broken and inconsistent, when result of an operation depends on presence of other independent operations. That brings the question why is such a half-done feature even needed.

Reply:
Let's move discussion of the issue to the issue.